### PR TITLE
LPS-151716 Adapt validations to accept DDM Expressions

### PR DESCRIPTION
**Pull request by seltong**

### CODE REVIEW

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:test:sf

File: modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/display/context/ObjectDefinitionsValidationsDisplayContext.java [hunk 1]
```diff
@@ -104,43 +157,145 @@ public List<Map<String, String>> getObjectValidationRuleEngines() {
...
 			"objectValidationRuleEngines", getObjectValidationRuleEngines()
 		).put(
 			"readOnly", !hasUpdateObjectDefinitionPermission()
 		).build();
 	}
 
+	public enum ValidationRuleElementFunctionItem {
+
+		CONCAT("concat(parameters)", "concat"),
+		CONTAINS("contains(field_reference, parameter)", "contains"),
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/display/context/ObjectDefinitionsValidationsDisplayContext.java:197]: @seltong, replace field_reference to field_name

File: modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/display/context/ObjectDefinitionsValidationsDisplayContext.java [hunk 2]
```diff
@@ -82,6 +84,57 @@ public List<FDSActionDropdownItem> getFDSActionDropdownItems()
 				"delete", "delete", "async"));
 	}
 
+	public List<HashMap<String, Object>> getObjectValidationRuleElements(
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/display/context/ObjectDefinitionsValidationsDisplayContext.java:87]: Make this method private

File: modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/display/context/ObjectDefinitionsValidationsDisplayContext.java [hunk 3]
```diff
@@ -104,43 +157,145 @@ public List<Map<String, String>> getObjectValidationRuleEngines() {
...
+					).put(
+						"tooltip", "placeholder"
+					).build());
+			}
+
+			functionItems.sort(
+				Comparator.comparing(
+					item -> item.get(
+						"label"
+					).toLowerCase()));
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/display/context/ObjectDefinitionsValidationsDisplayContext.java:240]: Remove this sort

**Review by MarinhoFeliphe — CHANGES_REQUESTED** [review]

File: modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/display/context/ObjectDefinitionsValidationsDisplayContext.java [hunk 4]
```diff
@@ -104,43 +157,145 @@ public List<Map<String, String>> getObjectValidationRuleEngines() {
...
+		private String _content;
+		private String _key;
+
+	}
+
+	public enum ValidationRuleElementOperatorItem {
+
+		AND("AND", "and"),
+		DIVIDED_BY("field_reference / field_reference2", "divided-by"),
+		EQUALS_OR_ATRIBUTE(
```
[the remark this answers is not in the document]
Comment (seltong) [reply modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/display/context/ObjectDefinitionsValidationsDisplayContext.java:259]: We are checking to see if there is a back-end logic for this.

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:test:relevant

**Comment by seltong** [comment]
ci:test:sf

**Comment by cgmonte** [comment]
Just started reviewing :)

File: modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/display/context/ObjectDefinitionsValidationsDisplayContext.java [hunk 5]
```diff
@@ -104,43 +108,131 @@ public List<Map<String, String>> getObjectValidationRuleEngines() {
...
+			"futureDates(field_name, parameter) AND pastDates(" +
+				"field_name, parameter)",
+			"range"),
+		SUM("sum(parameter)", "sum");
+
+		public static List<HashMap<String, String>> getValues(Locale locale) {
+			List<HashMap<String, String>> ddmExpressionFunctions =
+				new ArrayList<>();
+
+			for (DDMExpressionFunction functionItem : values()) {
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/display/context/ObjectDefinitionsValidationsDisplayContext.java:177]: rename to ddmExpressionFunction

File: modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/display/context/ObjectDefinitionsValidationsDisplayContext.java [hunk 6]
```diff
@@ -104,43 +108,131 @@ public List<Map<String, String>> getObjectValidationRuleEngines() {
...
+		AND("AND", "and"), DIVIDED_BY("field_name / field_name2", "divided-by"),
+		MINUS("field_name - field_name2", "minus"), OR("OR", "or"),
+		PLUS("field_name + field_name2", "plus"),
+		TIMES("field_name * field_name2", "times");
+
+		public static List<HashMap<String, String>> getValues(Locale locale) {
+			List<HashMap<String, String>> ddmExpressionOperators =
+				new ArrayList<>();
+
+			for (DDMExpressionOperator operatorItem : values()) {
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/display/context/ObjectDefinitionsValidationsDisplayContext.java:212]: rename to ddmExpressionOperator

File: modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/display/context/ObjectDefinitionsValidationsDisplayContext.java [hunk 7]
```diff
@@ -159,6 +251,52 @@ protected String getAPIURI() {
...
+					field -> HashMapBuilder.put(
+						"content", field.getName()
+					).put(
+						"label", field.getLabel(objectRequestHelper.getLocale())
+					).put(
+						"tooltip", StringPool.BLANK
+					).build()),
+				"fields"));
+
+		if (engine.equals("ddm")) {
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/display/context/ObjectDefinitionsValidationsDisplayContext.java:284]: Replace "ddm" to ObjectValidationRuleConstants.ENGINE_TYPE_DDM

**Review by MarinhoFeliphe — CHANGES_REQUESTED** [review]

**Comment by seltong** [comment]
ci:forward

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:forward

**Comment by seltong** [comment]
ci:stop

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:forward

**Comment by naraandrade** [comment]
ci:test:relevant

**Comment by seltong** [comment]
ci:forward

**Comment by seltong** [comment]
ci:stop

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:test:sf

**Comment by cgmonte** [comment]
Just started reviewing :)

**Comment by seltong** [comment]
ci:stop

**Comment by seltong** [comment]
ci:test:relevant

**Comment by seltong** [comment]
ci:stop

**Comment by seltong** [comment]
ci:forward

**Comment by seltong** [comment]
ci:test:relevant

**Comment by seltong** [comment]
ci:forward

**Comment by seltong** [comment]
ci:forward

**Comment by seltong** [comment]
ci:forward

**Comment by lucasperj** [comment]
Sent here: https://github.com/brianchandotcom/liferay-portal/pull/116852. Thank you.

**Pull request closed without merging**